File: cpp/questions/jump.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int ns = nums.size();
        int r = 0;
        int l = 0;
        int s = 0;
        while (l <= r) {
            if (r >= ns-1) {
                return s;
            }
            int cr = r;
            for (; l<=cr; ++l) {
                r = std::max(r, nums[l]+l);
            }
            ++s;
        }
        return -1;
    }
};

File: rust/questions/jump.rs
impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let nl = nums.len();
        let mut r: usize = 0;
        let mut l: usize = 0;
        let mut s = 0;
        while l <= r {
            if r >= nl - 1 {
                return s;
            }
            let cr = r;
            while l <= cr {
                r = r.max(nums[l] as usize + l);
                l += 1;
            }
            s += 1;
        }
        return -1;
    }
}
